eZ/Publish/Core/Persistence/Cache/InMemory/InMemoryCache.php

```diff
@@ -0,0 +1,192 @@
+            return null;
+        }
+
+        return $this->cache[$index];
+    }
+
+    /**
+     * Set object in in-memory cache.
+     *
+     * Should only set Cache hits here!
+     *
+     * @param object[] $objects
+     * @param callable $objectIndexes Return array of indexes per object (first argument), must return at least 1 primary index
+     * @param string|null $listIndex Optional index for list of items
+     */
+    public function setMulti(array $objects, callable $objectIndexes, string $listIndex = null): void
+    {
+        // If objects accounts for more then 20% of our limit, assume it's bulk content load and skip saving in-memory
+        if ($this->enabled === false || \count($objects) >= $this->limit / 5) {
+            return;
+        }
+
+        // check if we will reach limit by adding these objects, if so remove old cache
+        if (\count($this->cache) + \count($objects) >= $this->limit) {
+            $this->vacuum();
+        }
+
+        $time = microtime(true);
+        // if set add objects to cache on list index (typically a "all" key)
+        if ($listIndex) {
+            $this->cache[$listIndex] = $objects;
+            $this->cacheTime[$listIndex] = $time;
+        }
+
+        foreach ($objects as $object) {
+            // Skip if there are no indexes
+            if (!$indexes = $objectIndexes($object)) {
+                continue;
+            }
+
+            $key = array_shift($indexes);
+            $this->cache[$key] = $object;
+            $this->cacheTime[$key] = $time;
+
+            foreach ($indexes as $index) {
+                $this->cacheIndex[$index] = $key;
+            }
+        }
+    }
+
+    /**
+     * Removes multiple in-memory cache from the pool.
+     *
+     * @param string[] $keys An array of keys that should be removed from the pool.
+     */
+    public function deleteMulti(array $keys): void
+    {
+        if ($this->enabled === false) {
+            return;
+        }
+
+        foreach ($keys as $key) {
+            if ($index = $this->cacheIndex[$key] ?? null) {
+                unset($this->cacheIndex[$key], $this->cache[$index], $this->cacheTime[$index]);
+            } else {
+                unset($this->cache[$key], $this->cacheTime[$key]);
+            }
+        }
+    }
+
+    /**
+     * Deletes all cache in the in-memory pool.
+     */
+    public function clear(): void
+    {
+        // On purpose does not check if enabled, in case of several instances we allow clearing cache
+        $this->cache = $this->cacheIndex = $this->cacheTime = [];
+    }
+
+    /**
+     * Call to reduce cache items when $limit has been reached.
+     *
+     * Deletes expired first, then oldest(or least used?).
+     */
+    private function vacuum(): void
+    {
+        // Vacuuming cache in bulk, clearing the 33% oldest cache values
+        $this->cache = \array_slice($this->cache, (int) ($this->limit / 3));
```

Comment:
`$this->limit / 3` also could be calculated only once.

Reply:
For this and the / 5 comment:
1. I haven't seen this being called very often.
2. I don't expect the match here to take much ops compared to array_slice.